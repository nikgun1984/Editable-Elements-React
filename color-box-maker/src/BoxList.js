import NewBoxForm from "./NewBoxForm";
import Box from "./Box";
import { v4 as uuid } from "uuid";
import { useState } from "react";

const BoxList = () => {
	const [divs, setDivs] = useState([]);
	const addDiv = (div) => {
		let newDiv = { ...div, id: uuid() };
		setDivs((divs) => [...divs, newDiv]);
	};
	const removeDiv = (id) => {
		setDivs(divs.filter((div) => div.id !== id));
	};
	return (
		<div>
			<NewBoxForm addDiv={addDiv} />
			<div>
				{divs.map((div) => (
					<Box
						key={div.id}
						id={div.id}
						width={div.width}
						height={div.height}
						color={div.color}
						remove={removeDiv}
					/>
				))}
			</div>
		</div>
	);
};

export default BoxList;
