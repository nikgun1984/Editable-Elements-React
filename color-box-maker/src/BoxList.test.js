import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// smoke test
it("renders without crashing", function () {
	render(<BoxList />);
});

// snapshot test
it("matches snapshot", function () {
	const { asFragment } = render(<BoxList />);
	expect(asFragment()).toMatchSnapshot();
});

// specialized testing
it("can add a new div", function () {
	const { getByLabelText, queryByTestId, queryByText } = render(<BoxList />);

	//no todos just yet
	expect(queryByTestId("box-created")).not.toBeInTheDocument();

	const width = getByLabelText("Width");
	const height = getByLabelText("Height");
	const color = getByLabelText("Color");
	const btn = queryByText("Submit");

	// fill out the form
	fireEvent.change(width, { target: { value: "100" } });
	fireEvent.change(height, { target: { value: "100" } });
	fireEvent.change(color, { target: { value: "#ff0000" } });
	fireEvent.click(btn);

	expect(queryByTestId("box-created")).toBeInTheDocument();
	expect(queryByTestId("box-created")).toHaveClass("Box");
});

it("can remove a box", function () {
	const { getByLabelText, queryByTestId, queryByText } = render(<BoxList />);

	const width = getByLabelText("Width");
	const height = getByLabelText("Height");
	const color = getByLabelText("Color");
	const btn = queryByText("Submit");

	// fill out the form
	fireEvent.change(width, { target: { value: "100" } });
	fireEvent.change(height, { target: { value: "100" } });
	fireEvent.change(color, { target: { value: "#ff0000" } });
	fireEvent.click(btn);

	expect(queryByTestId("box-created")).toBeInTheDocument();

	const btnDelete = queryByText("X");
	fireEvent.click(btnDelete);
	expect(queryByTestId("box-created")).not.toBeInTheDocument();
});
