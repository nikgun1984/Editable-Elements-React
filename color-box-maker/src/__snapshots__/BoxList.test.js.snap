// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="width"
        >
          Width
        </label>
        <input
          id="width"
          name="width"
          placeholder="width"
          type="text"
          value="0"
        />
        <label
          for="height"
        >
          Height
        </label>
        <input
          id="height"
          name="height"
          placeholder="height"
          type="text"
          value="0"
        />
        <label
          for="color"
        >
          Color
        </label>
        <input
          id="color"
          name="color"
          placeholder="Background Color"
          type="color"
          value="#e3ae1e"
        />
        <button>
          Submit
        </button>
      </form>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
